packages/@uppy/dashboard/src/components/VirtualList.js

```diff
@@ -0,0 +1,136 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Adaptations:
+ * - Added role=presentation to helper elements
+ * - Tweaked styles for Uppy's Dashboard use case
+ */
+
+const { h, Component } = require('preact')
+
+const STYLE_INNER = {
+  position: 'relative',
+  // Disabled for our use case: the wrapper elements around FileList already deal with overflow,
+  // and this additional property would hide things that we want to show.
+  //
+  // overflow: 'hidden',
+  width: '100%',
+  minHeight: '100%'
+}
+
+const STYLE_CONTENT = {
+  position: 'absolute',
+  top: 0,
+  left: 0,
+  // Because the `top` value gets set to some offset, this `height` being 100% would make the scrollbar
+  // stretch far beyond the content. For our use case, the content div actually can get its height from
+  // the elements inside it, so we don't need to specify a `height` property at all.
+  //
+  // height: '100%',
+  width: '100%',
+  overflow: 'visible'
+}
+
+class VirtualList extends Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      offset: 0,
+      height: 0
+    }
+  }
+
+  resize = () => {
+    if (this.state.height !== this.base.offsetHeight) {
+      this.setState({
+        height: this.base.offsetHeight
+      })
+    }
+  }
+
+  handleScroll = () => {
+    this.setState({
+      offset: this.base.scrollTop
+    })
+    if (this.props.sync) {
+      this.forceUpdate()
+    }
+  }
+
+  componentDidUpdate () {
+    this.resize()
+  }
+
+  componentDidMount () {
+    this.resize()
+    addEventListener('resize', this.resize)
+  }
+
+  componentWillUnmount () {
+    removeEventListener('resize', this.resize)
+  }
+
+  render ({
+    data,
+    rowHeight,
+    renderRow,
+    overscanCount = 10,
+    sync,
+    ...props
+  }) {
+    const { offset, height } = this.state
+    // first visible row index
+    let start = Math.floor(offset / rowHeight)
+
+    // actual number of visible rows (without overscan)
+    let visibleRowCount = Math.floor(height / rowHeight)
+
+    // Overscan: render blocks of rows modulo an overscan row count
+    // This dramatically reduces DOM writes during scrolling
+    if (overscanCount) {
+      start = Math.max(0, start - (start % overscanCount))
+      visibleRowCount += overscanCount
+    }
+
+    // last visible + overscan row index
+    const end = start + 1 + visibleRowCount
+
+    // data slice currently in viewport plus overscan items
+    const selection = data.slice(start, end)
+
+    const styleInner = { ...STYLE_INNER, height: data.length * rowHeight }
+    const styleContent = { ...STYLE_CONTENT, top: start * rowHeight }
+
+    return (
+      <div onScroll={this.handleScroll} {...props}>
+        <div role="presentation" style={styleInner}>
+          <div role="presentation" style={styleContent}>
```

Comment:
Divs already have no semantics, I believe we don't need to strip them of their semantics with `role="presentation"`.
Also - since we inlined VirtualList (good idea), let's change `<div role="listitem">` back to `<li>`?

Reply:
I think the `role="presentation"` is required to make the `role="listitem"` elements act as direct children of the `role="list"` root element. StackOverflow told me a `role="list"` element normally assumes its children are list items, unless they're marked with `role="presentation"`.

We could change the innermost `<div>` to a `<ul>` and then use `<li>` for the items again, though, I think…

Reply:
Oh! But agreed, let's just use `<ul>`.
I wonder how accessible virtual lists are in general, does it give us a significant performance improvement, have you checked?

Reply:
Also we can't tab through the virtual list properly :-o